File: frontend/src/query.test.ts
import { describe, expect, it } from "vitest";
import formatMetadata from "../../document/format_metadata.json";
import fixture from "../../internal/query/testdata/identity-vectors.json";
import {
  canonicalHighlightSet,
  canonicalQuery,
  classifyMedia,
  highlightSetFingerprint,
  parseHighlightSet,
  parseQuery,
  queryFingerprint,
} from "./query.js";

describe("query identity", () => {
  for (const vector of fixture.queries) {
    it(vector.name, async () => {
      const value = parseQuery(vector.input_json);
      expect(canonicalQuery(value)).toBe(vector.canonical_utf8);
      expect(await queryFingerprint(value)).toBe(vector.fingerprint);
    });
  }

  for (const testCase of fixture.invalid_queries) {
    it(`rejects ${testCase.name}`, () => {
      expect(() => parseQuery(testCase.json_text)).toThrow();
    });
  }

  it("accepts null only for optional filters", () => {
    expect(canonicalQuery(parseQuery('{"filters":{"paths":null,"size_min":null,"no_tags":true}}')))
      .toBe('{"filters":{"no_tags":true},"mode":"lexical","sort":{"direction":"asc","field":"name"},"syntax":"simple","text":"","v":1}');
    expect(() => parseQuery('{"sort":null}')).toThrow();
  });

  it("accepts a valid escaped astral scalar", () => {
    expect(parseQuery('{"text":"\\ud83d\\ude00"}').text).toBe("😀");
  });

  it("rejects invalid values and bounds", () => {
    const invalid = [
      '{"filters":{"paths":["relative"]}}',
      '{"filters":{"tag_ids":["00000000-0000-3000-8000-000000000001"]}}',
      '{"filters":{"mime_types":["text/*"]}}',
      '{"filters":{"extensions":[".pdf"]}}',
      '{"v":0}',
      '{"syntax":""}',
      '{"sort":{"field":""}}',
      '{"filters":{"modified_after":""}}',
      '{"filters":{"modified_after":"2026-09-10T00:00:00Z","modified_before":"2026-09-09T00:00:00Z"}}',
      '{"filters":{"modified_after":"2026-09-09T00:00:00.1Z","modified_before":"2026-09-09T00:00:00Z"}}',
      '{"filters":{"size_min":2,"size_max":1}}',
      `{"text":"${"😀".repeat(8193)}"}`,
      `${" ".repeat(128 * 1024)}{}`,
    ];
    for (const raw of invalid) expect(() => parseQuery(raw)).toThrow();
    expect(() => parseQuery('{"filters":{"modified_after":"2026-09-09T00:00:00Z","modified_before":"2026-09-09T00:00:00.1Z"}}')).not.toThrow();
  });

  it("enforces the shared normalized query byte bound during parse", () => {
    const raw = `{"filters":{"paths":["/${"x".repeat(fixture.bounds.oversized_query_path_ascii_bytes)}"]}}`;
    expect(new TextEncoder().encode(raw).byteLength).toBeLessThan(128 * 1024);
    expect(() => parseQuery(raw)).toThrow();
  });
});

describe("highlight-set identity", () => {
  for (const vector of fixture.highlight_sets) {
    it(vector.name, async () => {
      const value = parseHighlightSet(vector.input_json);
      expect(canonicalHighlightSet(value)).toBe(vector.canonical_utf8);
      expect(await highlightSetFingerprint(value)).toBe(vector.fingerprint);
    });
  }

  for (const testCase of fixture.invalid_highlight_sets) {
    it(`rejects ${testCase.name}`, () => {
      expect(() => parseHighlightSet(testCase.json_text)).toThrow();
    });
  }

  it("rejects query payloads, duplicates, unknowns, and bounds", () => {
    const invalid = [
      '{"v":1,"text":"query"}',
      '{"v":1,"terms":[]}',
      '{"v":1,"terms":[{"text":"same","color":"#ffffff"},{"text":"same","color":"#000000"}]}',
      '{"v":1,"terms":[{"text":"term","color":"#ABCDEF"}]}',
      '{"v":1,"terms":[{"text":"term","color":"#abcdef","regex":true}]}',
      `{"v":1,"terms":[{"text":"${"😀".repeat(257)}","color":"#abcdef"}]}`,
    ];
    for (const raw of invalid) expect(() => parseHighlightSet(raw)).toThrow();
  });

  it("enforces shared highlight boundaries during parse", () => {
    const maxTerm = `{"v":1,"terms":[{"text":"${"x".repeat(fixture.bounds.highlight_max_term_scalars)}","color":"#abcdef"}]}`;
    expect(() => parseHighlightSet(maxTerm)).not.toThrow();
    const maxTerms = Array.from({ length: fixture.bounds.highlight_max_terms }, (_, index) =>
      `{"text":"${String(index).padStart(2, "0")}","color":"#abcdef"}`);
    expect(() => parseHighlightSet(`{"v":1,"terms":[${maxTerms.join(",")}]}`)).not.toThrow();

    const oversizedTerms = Array.from({ length: fixture.bounds.highlight_max_terms }, (_, index) =>
      `{"text":"${"😀".repeat(fixture.bounds.oversized_highlight_astral_scalars)}${String(index).padStart(2, "0")}","color":"#abcdef"}`);
    const oversized = `{"v":1,"terms":[${oversizedTerms.join(",")}]}`;
    expect(new TextEncoder().encode(oversized).byteLength).toBeLessThan(128 * 1024);
    expect(() => parseHighlightSet(oversized)).toThrow();
  });
});

describe("media classification", () => {
  for (const testCase of fixture.media_cases) {
    it(testCase.name, () => {
      expect(classifyMedia(testCase.media_type, testCase.filename)).toBe(testCase.family);
    });
  }

  it("uses every document-owned mapping", () => {
    for (const metadata of formatMetadata) {
      expect(classifyMedia(metadata.media_type, "synthetic.bin"), metadata.id).toBe(metadata.query_family);
      for (const extension of metadata.extensions) {
        expect(classifyMedia("application/octet-stream", `synthetic.${extension}`), metadata.id).toBe(metadata.query_family);
      }
    }
  });

  it.each([
    ["message/rfc822", "synthetic.eml", "email"],
    ["application/pdf", "synthetic.pdf", "document"],
    ["text/csv", "synthetic.csv", "spreadsheet"],
    ["application/vnd.ms-powerpoint", "synthetic.ppt", "presentation"],
    ["image/vnd.synthetic", "synthetic.bin", "image"],
    ["video/vnd.synthetic", "synthetic.bin", "audio_video"],
    ["text/plain", "synthetic.txt", "text"],
    ["text/x-python", "synthetic.py", "source_code"],
    ["text/html", "synthetic.html", "web"],
    ["text/calendar", "synthetic.ics", "calendar"],
    ["application/zip", "synthetic.zip", "archive"],
    ["image/vnd.dwg", "synthetic.dwg", "cad"],
    ["application/vnd.synthetic-unknown", "synthetic.pdf", "unknown"],
  ])("classifies %s as %s", (mediaType, filename, family) => {
    expect(classifyMedia(mediaType, filename)).toBe(family);
  });

  it("uses extension only for missing or generic MIME", () => {
    expect(classifyMedia("application/octet-stream", "REPORT.PDF")).toBe("document");
    expect(classifyMedia("application/octet-stream", "scan.PNG")).toBe("image");
    expect(classifyMedia("application/octet-stream", "recording.MP3")).toBe("audio_video");
    expect(classifyMedia("", "main.GO")).toBe("source_code");
    expect(classifyMedia("application/vnd.synthetic-unknown", "report.pdf")).toBe("unknown");
  });
});

File: internal/query/query_test.go
package query

import (
	_ "embed"
	"encoding/json/v2"
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

//go:embed testdata/identity-vectors.json
var identityFixtureJSON []byte

type identityFixture struct {
	Format               string           `json:"format"`
	Queries              []identityVector `json:"queries"`
	HighlightSets        []identityVector `json:"highlight_sets"`
	InvalidQueries       []invalidVector  `json:"invalid_queries"`
	InvalidHighlightSets []invalidVector  `json:"invalid_highlight_sets"`
	Bounds               boundaryFixture  `json:"bounds"`
	MediaCases           []mediaCase      `json:"media_cases"`
}

type identityVector struct {
	Name          string `json:"name"`
	InputJSON     string `json:"input_json"`
	CanonicalUTF8 string `json:"canonical_utf8"`
	Fingerprint   string `json:"fingerprint"`
}

type invalidVector struct {
	Name     string `json:"name"`
	JSONText string `json:"json_text"`
}

type boundaryFixture struct {
	OversizedQueryPathASCIIBytes    int `json:"oversized_query_path_ascii_bytes"`
	HighlightMaxTerms               int `json:"highlight_max_terms"`
	HighlightMaxTermScalars         int `json:"highlight_max_term_scalars"`
	OversizedHighlightAstralScalars int `json:"oversized_highlight_astral_scalars"`
}

type mediaCase struct {
	Name      string `json:"name"`
	MediaType string `json:"media_type"`
	Filename  string `json:"filename"`
	Family    string `json:"family"`
}

func loadIdentityFixture(t *testing.T) identityFixture {
	t.Helper()
	var fixture identityFixture
	require.NoError(t, json.Unmarshal(identityFixtureJSON, &fixture, json.RejectUnknownMembers(true)))
	require.Equal(t, "docbank-query-identity-v1", fixture.Format)
	return fixture
}

func TestQueryIdentityVectors(t *testing.T) {
	for _, vector := range loadIdentityFixture(t).Queries {
		t.Run(vector.Name, func(t *testing.T) {
			value, err := Parse([]byte(vector.InputJSON))
			require.NoError(t, err)
			encoded, fingerprint, err := CanonicalWithFingerprint(value)
			require.NoError(t, err)
			assert.Equal(t, vector.CanonicalUTF8, string(encoded))
			assert.Equal(t, vector.Fingerprint, fingerprint)
		})
	}
}

func TestQueryRejectsInvalidSharedInputs(t *testing.T) {
	for _, testCase := range loadIdentityFixture(t).InvalidQueries {
		t.Run(testCase.Name, func(t *testing.T) {
			_, err := Parse([]byte(testCase.JSONText))
			require.Error(t, err)
		})
	}
}

func TestQueryNormalizesOptionalNullsAndFlags(t *testing.T) {
	value, err := Parse([]byte(`{"filters":{"paths":null,"modified_after":null,"size_min":null,"no_tags":true,"exclude_tag_ids":["00000000-0000-4000-8000-000000000001"]}}`))
	require.NoError(t, err)
	encoded, err := Canonical(value)
	require.NoError(t, err)
	assert.Equal(t, `{"filters":{"exclude_tag_ids":["00000000-0000-4000-8000-000000000001"],"no_tags":true},"mode":"lexical","sort":{"direction":"asc","field":"name"},"syntax":"simple","text":"","v":1}`, string(encoded)) //nolint:testifylint // Exact bytes are the contract.

	_, err = Parse([]byte(`{"filters":{"no_tags":true,"tag_ids":["00000000-0000-4000-8000-000000000001"]}}`))
	require.Error(t, err)
}

func TestQueryAcceptsEscapedAstralUnicode(t *testing.T) {
	value, err := Parse([]byte(`{"text":"\ud83d\ude00"}`))
	require.NoError(t, err)
	assert.Equal(t, "😀", value.Text)
}

func TestQueryRejectsInvalidFilterValues(t *testing.T) {
	tests := map[string]string{
		"relative path":             `{"filters":{"paths":["relative"]}}`,
		"dot path segment":          `{"filters":{"paths":["/one/../two"]}}`,
		"root trailing slash":       `{"filters":{"paths":["/one/"]}}`,
		"wrong UUID version":        `{"filters":{"tag_ids":["00000000-0000-3000-8000-000000000001"]}}`,
		"uppercase UUID":            `{"filters":{"tag_ids":["00000000-0000-4000-8000-00000000000A"]}}`,
		"unknown media family":      `{"filters":{"media_families":["future"]}}`,
		"parameterized MIME":        `{"filters":{"mime_types":["text/plain; charset=utf-8"]}}`,
		"wildcard MIME":             `{"filters":{"mime_types":["text/*"]}}`,
		"uppercase MIME":            `{"filters":{"mime_types":["Text/Plain"]}}`,
		"leading-dot extension":     `{"filters":{"extensions":[".pdf"]}}`,
		"unknown coverage":          `{"filters":{"text_coverage":["pending"]}}`,
		"zero version":              `{"v":0}`,
		"empty syntax":              `{"syntax":""}`,
		"empty sort field":          `{"sort":{"field":""}}`,
		"empty timestamp":           `{"filters":{"modified_after":""}}`,
		"inverted times":            `{"filters":{"modified_after":"2026-09-10T00:00:00Z","modified_before":"2026-09-09T00:00:00Z"}}`,
		"equal times":               `{"filters":{"modified_after":"2026-09-09T00:00:00Z","modified_before":"2026-09-09T00:00:00Z"}}`,
		"inverted sizes":            `{"filters":{"size_min":2,"size_max":1}}`,
		"negative size":             `{"filters":{"size_min":-1}}`,
		"too many supplied entries": `{"filters":{"extensions":[` + strings.Repeat(`"txt",`, 32) + `"txt"]}}`,
	}
	for name, raw := range tests {
		t.Run(name, func(t *testing.T) {
			_, err := Parse([]byte(raw))
			require.Error(t, err)
		})
	}
}

func TestQueryEnforcesInputTextAndDepthBounds(t *testing.T) {
	fixture := loadIdentityFixture(t)
	_, err := Parse([]byte(strings.Repeat(" ", 128<<10) + `{}`))
	require.Error(t, err)

	_, err = Parse([]byte(`{"text":"` + strings.Repeat("😀", 8193) + `"}`))
	require.Error(t, err)

	_, err = Parse([]byte(`{"filters":{"paths":["/` + strings.Repeat("x", fixture.Bounds.OversizedQueryPathASCIIBytes) + `"]}}`))
	require.Error(t, err)

	_, err = Parse([]byte(`{"filters":{"paths":[[[[[[[[[[[[[[[[["/deep"]]]]]]]]]]]]]]]]]}}`))
	require.Error(t, err)
}

func TestCanonicalRejectsInvalidProgrammaticQuery(t *testing.T) {
	_, err := Canonical(Query{V: 1, Text: string([]byte{0xff})})
	require.Error(t, err)
}
